SingularityUI/app/components/requestForm/RequestForm.jsx

```diff
@@ -0,0 +1,560 @@
+    } else if (this.isEditing() && this.props.request.request[fieldId] !== undefined) {
+      return this.props.request.request[fieldId];
+    } else {
+      return ""
+    }
+  }
+
+  cantSubmit() {
+    return this.props.saveApiCall.isFetching ||
+      !this.getValue('id') ||
+      !this.getValue('requestType') ||
+      (this.getValue('requestType') === 'SCHEDULED' && !this.getValue(this.getScheduleType()));
+  }
+
+  submitForm(event) {
+    event.preventDefault();
+    const request = {};
+    const copyOverField = (fieldId) => {
+      if (this.getValue(fieldId) && fieldId != QUARTZ_SCHEDULE && fieldId != CRON_SCHEDULE && fieldId != 'scheduleType') {
+        request[fieldId] = this.getValue(fieldId);
+      }
+    }
+    
+    FIELDS_BY_REQUEST_TYPE[this.getValue('requestType')].map(copyOverField)
+    FIELDS_BY_REQUEST_TYPE.ALL.map(copyOverField)
+
+    if (this.getValue('requestType') === 'SCHEDULED') {
+      if (this.getScheduleType() === QUARTZ_SCHEDULE) {
+        request[QUARTZ_SCHEDULE] = this.getValue(QUARTZ_SCHEDULE);
+      } else {
+        request.schedule = this.getValue(CRON_SCHEDULE);
+      }
+    }
+
+    if (['ON_DEMAND', 'RUN_ONCE'].indexOf(this.getValue('requestType')) !== -1) {
+      request.daemon = false
+    } else if (['SERVICE', 'WORKER'].indexOf(this.getValue('requestType')) !== -1) {
+      request.daemon = true
+    }
+
+    if (request.owners) {
+      request.owners = request.owners.split(',')
+    }
+    if (request.rackAffinity) {
+      request.rackAffinity = request.rackAffinity.split(',')
+    }
+
+    this.props.save(request);
+    return null;
+  }
+
+  shouldRenderField(fieldId) {
+    if (!this.getValue('requestType')) {
+      return false;
+    } else if (FIELDS_BY_REQUEST_TYPE[this.getValue('requestType')].indexOf(fieldId) === -1) {
+      return false;
+    } else {
+      return true;
+    }
+  }
+
+  getButtonsDisabled(type) {
+    if (this.isEditing() && this.getValue('requestType') !== type) {
+      return 'disabled';
+    } else {
+      return null;
+    }
+  }
+
+  updateField(fieldId, newValue) {
+    this.props.update(FORM_ID, fieldId, newValue)
+  }
+
+  updateTypeButtonClick(event) {
+    event.preventDefault();
+    this.updateField('requestType', event.target.value);
+  }
+
+  getScheduleType() {
+    if (this.isEditing() && !(this.props.form && this.props.form.scheduleType)) {
+      if (this.props.request.request[QUARTZ_SCHEDULE]) {
+        return QUARTZ_SCHEDULE;
+      } else {
+        return CRON_SCHEDULE;
+      }
+    } else {
+      if (this.props.form && this.props.form.scheduleType) {
+        return this.props.form.scheduleType;
+      } else {
+        return CRON_SCHEDULE;
+      }
+    }
+  }
+
+  renderRequestTypeSelectors() {
+    const selectors = [];
+    const tooltip = (
+      <ToolTip id="cannotChangeRequestTypeAfterCreation">Option cannot be altered after creation</ToolTip>
+    );
+    REQUEST_TYPES.map((requestType, key) => {
```

Comment:
`map()` returns an array of all the return values of the inner method -- it'd be much cleaner to do something like:

```javascript
const selectors = REQUEST_TYPES.map((requestType, key) => { return <TheThingToBeAdded/>; })
```